src/api/utils.rs

```diff
@@ -19,4 +25,110 @@ pub mod utils {
         let client = crate::client::get_client();
         client.utils().is_steam_running_on_steam_deck()
     }
+
+    #[napi]
+    pub enum GamepadTextInputMode {
+        Normal,
+        Password,
+    }
+
+    #[napi]
+    pub enum GamepadTextInputLineMode {
+        SingleLine,
+        MultipleLines,
+    }
+
+    /// @returns the entered text, or null if cancelled or could not show the input
+    #[napi]
+    pub async fn show_gamepad_text_input(
+        input_mode: GamepadTextInputMode,
+        input_line_mode: GamepadTextInputLineMode,
+        description: String,
+        max_characters: u32,
+        existing_text: Option<String>,
+    ) -> Option<String> {
+        let client = crate::client::get_client();
+
+        let (tx, rx) = oneshot::channel();
+        let mut tx = Some(tx);
+
+        let opened = client.utils().show_gamepad_text_input(
+            match input_mode {
+                GamepadTextInputMode::Normal => kGamepadTextInputMode::Normal,
+                GamepadTextInputMode::Password => kGamepadTextInputMode::Password,
+            },
+            match input_line_mode {
+                GamepadTextInputLineMode::SingleLine => kGamepadTextInputLineMode::SingleLine,
+                GamepadTextInputLineMode::MultipleLines => kGamepadTextInputLineMode::MultipleLines,
+            },
+            &description,
+            max_characters,
+            existing_text.as_deref(),
+            move |dismissed_data| {
+                if let Some(tx) = tx.take() {
+                    let text = client
+                        .utils()
+                        .get_entered_gamepad_text_input(&dismissed_data);
+                    tx.send(text).unwrap();
+                }
```

Comment:
What if there is no result here? The Promise will never resolve, right?

Looking further at the [docs](https://partner.steamgames.com/doc/api/ISteamUtils#ShowGamepadTextInput), this function returns true if "big picture overlay is running" but I'm wondering if that's a docs mistake and it returns true/false if the keyboard was opened or not similar to the floating text input function (or if it kinda means the same in this case, anyway).

Since the two new functions can be called without showing the gamepad, there is one more wrinkle in the Promise based API: Technically, promises are expected to be resolved at some point in the future, but if the gamepad doesn't open or doesn't exist, the Promise will never be handled. This is kinda ok-ish when using the Promise API, but it's very confusing when using async/await since the code block will just get stuck:

```js
try {
  const result = await showGamepadTextInput(…);
  doSomething(result); // This will never be called.
} catch (error) {
  // This also never happens.
}
```

Most likely this is fine if it's documented. One thing to keep in mind is that this behavior is likely attached to input elements and is called repeatedly. Without knowing whether the gamepad can/is ever shown on the platform that it is being called on it will create a lot of dangling promises in memory that will stay unresolved for the duration of the app runtime.

Now, this might be an acceptable trade-off for you, but I wanted to bring it up anyway for consideration.

Reply:
> What if there is no result here? The Promise will never resolve, right? (and the questions below that)

I don't think so, the memory management will drop the promise and unsign the callback as soon as the function returns something.

The idea is to always resolve the promise, either with the text input or with an undefined state, meaning the text input could not be shown or was cancelled. Example impl:

```js
const input = await showGamepadTextInput(…);
if (input) {
    doSomething(input);
} else {
    // do nothing, as use cancelled the input request
}
```

> Looking further at the [docs](https://partner.steamgames.com/doc/api/ISteamUtils#ShowGamepadTextInput), this function returns true if "big picture overlay is running" but I'm wondering if that's a docs mistake and it returns true/false if the keyboard was opened or not similar to the floating text input function (or if it kinda means the same in this case, anyway).

I think it means the same thing 🤔